File: .agents/skills/qa_tester/scripts/runner_adapter.py
#!/usr/bin/env python3
"""
runner_adapter.py — Adaptador de execução de testes para qa_tester no projeto MIST.

Funcionalidades:
1. Parseia TESTS.md para extrair testes e seus metadados.
2. Executa testes reais via subprocess para pytest, vitest e playwright.
3. Normaliza a saída de cada runner para o schema comum do resultados.json.
4. Aplica escrita atômica (arquivo temporário + rename).
5. Mantém histórico FIFO das últimas 10 execuções por ID de teste.
"""

import sys
import os
import re
import json
import time
import tempfile
import datetime
import subprocess
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[3]
TESTS_MD_PATH = ROOT_DIR / "TESTS.md"
RESULTADOS_JSON_PATH = ROOT_DIR / "resultados.json"


def parse_tests_md(category_filter=None, test_id_filter=None):
    if not TESTS_MD_PATH.exists():
        raise FileNotFoundError(f"Arquivo não encontrado: {TESTS_MD_PATH}")

    content = TESTS_MD_PATH.read_text(encoding="utf-8")
    lines = content.splitlines()

    tests = []
    current_category = None
    current_feature = None
    current_test = None

    category_regex = re.compile(r"^##\s+(?!Runners registrados)(.+)$")
    feature_regex = re.compile(r"^###\s+(.+)$")
    test_header_regex = re.compile(r"^####\s+([A-Za-z0-9_\-]+)\s+—\s+(.+)$")

    for line in lines:
        cat_match = category_regex.match(line)
        if cat_match:
            current_category = cat_match.group(1).strip()
            continue

        feat_match = feature_regex.match(line)
        if feat_match:
            current_feature = feat_match.group(1).strip()
            continue

        test_match = test_header_regex.match(line)
        if test_match:
            if current_test:
                tests.append(current_test)
            test_id = test_match.group(1).strip()
            title = test_match.group(2).strip()
            current_test = {
                "id": test_id,
                "title": title,
                "categoria": current_category,
                "funcionalidade": current_feature,
                "prioridade": "P1",
                "status": "planejado",
                "runner": "pytest",
                "comando": ""
            }
            continue

        if current_test:
            if line.strip().startswith("- Prioridade:"):
                current_test["prioridade"] = line.split(":", 1)[1].strip()
            elif line.strip().startswith("- Status:"):
                current_test["status"] = line.split(":", 1)[1].strip()
            elif line.strip().startswith("- Runner:"):
                current_test["runner"] = line.split(":", 1)[1].strip()
            elif line.strip().startswith("- Comando:"):
                cmd_raw = line.split(":", 1)[1].strip()
                current_test["comando"] = cmd_raw.strip("`").strip()

    if current_test:
        tests.append(current_test)

    # Filtragem
    filtered = []
    for t in tests:
        if test_id_filter and t["id"].lower() != test_id_filter.lower():
            continue
        if category_filter and t["categoria"].lower() != category_filter.lower():
            continue
        filtered.append(t)

    return filtered


def run_command(command, runner_name, cwd=ROOT_DIR):
    start_time = time.perf_counter()
    try:
        proc = subprocess.run(
            command,
            shell=True,
            cwd=cwd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace"
        )
        duration_ms = int((time.perf_counter() - start_time) * 1000)
        success = (proc.returncode == 0)
        output = proc.stdout + "\n" + proc.stderr
        error_msg = None if success else (proc.stderr.strip() or proc.stdout.strip() or f"Process returned exit code {proc.returncode}")
        return {
            "status": "pass" if success else "fail",
            "duracao_ms": duration_ms,
            "erro": error_msg
        }
    except Exception as exc:
        duration_ms = int((time.perf_counter() - start_time) * 1000)
        return {
            "status": "fail",
            "duracao_ms": duration_ms,
            "erro": str(exc)
        }


def load_resultados():
    if not RESULTADOS_JSON_PATH.exists():
        return {}
    try:
        content = RESULTADOS_JSON_PATH.read_text(encoding="utf-8").strip()
        if not content:
            return {}
        return json.loads(content)
    except Exception:
        return {}


def atomic_save_resultados(data):
    parent_dir = RESULTADOS_JSON_PATH.parent
    fd, temp_file_path = tempfile.mkstemp(dir=parent_dir, prefix="res_", suffix=".tmp")
    with os.fdopen(fd, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
        f.write("\n")

    # Substituição atômica no SO
    os.replace(temp_file_path, RESULTADOS_JSON_PATH)


def execute_suite(category=None, test_id=None, origem="execucao"):
    tests = parse_tests_md(category_filter=category, test_id_filter=test_id)
    if not tests:
        print(f"[qa_tester] Nenhum teste encontrado com os critérios (categoria='{category}', id='{test_id}').")
        return

    resultados = load_resultados()
    run_timestamp = datetime.datetime.now(datetime.timezone.utc).isoformat()

    print(f"[qa_tester] Iniciando processamento de {len(tests)} teste(s) [origem={origem}]...")

    for test in tests:
        t_id = test["id"]
        t_runner = test["runner"]
        t_status = test["status"]
        t_cmd = test["comando"]

        if t_id not in resultados:
            resultados[t_id] = {
                "runner": t_runner,
                "historico": []
            }
        else:
            resultados[t_id]["runner"] = t_runner

        if t_status == "planejado" or not t_cmd:
            entry = {
                "run_id": run_timestamp,
                "status": "pending",
                "duracao_ms": 0,
                "erro": None,
                "origem": origem
            }
            print(f" - [{t_id}] PENDING (planejado ou sem comando automatizado)")
        else:
            print(f" - [{t_id}] EXECUTANDO ({t_runner}): `{t_cmd}`")
            exec_res = run_command(t_cmd, t_runner)
            entry = {
                "run_id": run_timestamp,
                "status": exec_res["status"],
                "duracao_ms": exec_res["duracao_ms"],
                "erro": exec_res["erro"],
                "origem": origem
            }
            print(f"   └─ Resultado: {entry['status'].upper()} em {entry['duracao_ms']}ms")

        # Manter histórico FIFO das últimas 10 execuções
        historico = resultados[t_id].get("historico", [])
        historico.append(entry)
        if len(historico) > 10:
            historico = historico[-10:]
        resultados[t_id]["historico"] = historico

    # Escrita atômica garantida
    atomic_save_resultados(resultados)
    print(f"[qa_tester] Resultados salvos atomicamente em {RESULTADOS_JSON_PATH}")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Executor de testes para a skill qa_tester")
    parser.add_argument("--categoria", help="Filtrar por categoria do TESTS.md")
    parser.add_argument("--id", help="Filtrar por ID de teste único")
    parser.add_argument("--origem", default="execucao", choices=["execucao", "validacao"], help="Origem da execução")

    args = parser.parse_args()
    execute_suite(category=args.categoria, test_id=args.id, origem=args.origem)

File: .agents/skills/qa_tester/scripts/runner_adapter.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const ROOT_DIR = path.resolve(__dirname, '../../../..');
const TESTS_MD_PATH = path.join(ROOT_DIR, 'TESTS.md');
const RESULTADOS_JSON_PATH = path.join(ROOT_DIR, 'resultados.json');

function parseTestsMd(categoryFilter, testIdFilter) {
  if (!fs.existsSync(TESTS_MD_PATH)) {
    throw new Error(`Arquivo não encontrado: ${TESTS_MD_PATH}`);
  }

  const content = fs.readFileSync(TESTS_MD_PATH, 'utf-8');
  const lines = content.split(/\r?\n/);

  const tests = [];
  let currentCategory = null;
  let currentFeature = null;
  let currentTest = null;

  const categoryRegex = /^##\s+(?!Runners registrados)(.+)$/;
  const featureRegex = /^###\s+(.+)$/;
  const testHeaderRegex = /^####\s+([A-Za-z0-9_\-]+)\s+—\s+(.+)$/;

  for (const line of lines) {
    const catMatch = line.match(categoryRegex);
    if (catMatch) {
      currentCategory = catMatch[1].trim();
      continue;
    }

    const featMatch = line.match(featureRegex);
    if (featMatch) {
      currentFeature = featMatch[1].trim();
      continue;
    }

    const testMatch = line.match(testHeaderRegex);
    if (testMatch) {
      if (currentTest) tests.push(currentTest);
      currentTest = {
        id: testMatch[1].trim(),
        title: testMatch[2].trim(),
        categoria: currentCategory,
        funcionalidade: currentFeature,
        prioridade: 'P1',
        status: 'planejado',
        runner: 'pytest',
        comando: ''
      };
      continue;
    }

    if (currentTest) {
      const trimmed = line.trim();
      if (trimmed.startsWith('- Prioridade:')) {
        currentTest.prioridade = trimmed.split(':', 2)[1].trim();
      } else if (trimmed.startsWith('- Status:')) {
        currentTest.status = trimmed.split(':', 2)[1].trim();
      } else if (trimmed.startsWith('- Runner:')) {
        currentTest.runner = trimmed.split(':', 2)[1].trim();
      } else if (trimmed.startsWith('- Comando:')) {
        const rawCmd = trimmed.substring(trimmed.indexOf(':') + 1).trim();
        currentTest.comando = rawCmd.replace(/^`|`$/g, '').trim();
      }
    }
  }

  if (currentTest) tests.push(currentTest);

  return tests.filter(t => {
    if (testIdFilter && t.id.toLowerCase() !== testIdFilter.toLowerCase()) return false;
    if (categoryFilter && t.categoria.toLowerCase() !== categoryFilter.toLowerCase()) return false;
    return true;
  });
}

function runCommand(command, cwd = ROOT_DIR) {
  const startTime = Date.now();
  try {
    const output = execSync(command, { cwd, encoding: 'utf-8', stdio: 'pipe' });
    const durationMs = Date.now() - startTime;
    return { status: 'pass', duracao_ms: durationMs, erro: null };
  } catch (error) {
    const durationMs = Date.now() - startTime;
    const errorMsg = (error.stderr || error.stdout || error.message || 'Process failed').trim();
    return { status: 'fail', duracao_ms: durationMs, erro: errorMsg };
  }
}

function loadResultados() {
  if (!fs.existsSync(RESULTADOS_JSON_PATH)) return {};
  try {
    const content = fs.readFileSync(RESULTADOS_JSON_PATH, 'utf-8').trim();
    return content ? JSON.parse(content) : {};
  } catch (err) {
    return {};
  }
}

function atomicSaveResultados(data) {
  const tempPath = path.join(ROOT_DIR, `res_${Date.now()}_${Math.random().toString(36).substr(2, 6)}.tmp`);
  fs.writeFileSync(tempPath, JSON.stringify(data, null, 2) + '\n', 'utf-8');
  fs.renameSync(tempPath, RESULTADOS_JSON_PATH);
}

function executeSuite({ category, testId, origem = 'execucao' }) {
  const tests = parseTestsMd(category, testId);
  if (tests.length === 0) {
    console.log(`[qa_tester] Nenhum teste encontrado com os critérios (categoria='${category}', id='${testId}').`);
    return;
  }

  const resultados = loadResultados();
  const runTimestamp = new Date().toISOString();

  console.log(`[qa_tester] Iniciando processamento de ${tests.length} teste(s) [origem=${origem}]...`);

  for (const test of tests) {
    const { id: t_id, runner: t_runner, status: t_status, comando: t_cmd } = test;

    if (!resultados[t_id]) {
      resultados[t_id] = { runner: t_runner, historico: [] };
    } else {
      resultados[t_id].runner = t_runner;
    }

    let entry;
    if (t_status === 'planejado' || !t_cmd) {
      entry = {
        run_id: runTimestamp,
        status: 'pending',
        duracao_ms: 0,
        erro: null,
        origem
      };
      console.log(` - [${t_id}] PENDING (planejado ou sem comando automatizado)`);
    } else {
      console.log(` - [${t_id}] EXECUTANDO (${t_runner}): \`${t_cmd}\``);
      const execRes = runCommand(t_cmd);
      entry = {
        run_id: runTimestamp,
        status: execRes.status,
        duracao_ms: execRes.duracao_ms,
        erro: execRes.erro,
        origem
      };
      console.log(`   └─ Resultado: ${entry.status.toUpperCase()} em ${entry.duracao_ms}ms`);
    }

    let historico = resultados[t_id].historico || [];
    historico.push(entry);
    if (historico.length > 5) {
      historico = historico.slice(-5);
    }
    resultados[t_id].historico = historico;
  }

  atomicSaveResultados(resultados);
  console.log(`[qa_tester] Resultados salvos atomicamente em ${RESULTADOS_JSON_PATH}`);
}

const args = process.argv.slice(2);
let category = null;
let testId = null;
let origem = 'execucao';

for (let i = 0; i < args.length; i++) {
  if (args[i] === '--categoria' && args[i + 1]) category = args[++i];
  if (args[i] === '--id' && args[i + 1]) testId = args[++i];
  if (args[i] === '--origem' && args[i + 1]) origem = args[++i];
}

executeSuite({ category, testId, origem });
